From e78d1e91a6e66757482138da54c4d2efa2e668ff Mon Sep 17 00:00:00 2001
From: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
Date: Fri, 29 May 2020 13:39:10 +0300
Subject: [PATCH 192/245] mount: use ns_mountpoint in root_path_from_parent

Fail root_path_from_parent if parent is root_yard, we want to only
lookup root path in real parent mounts.

Now it is safe to use ns_mountpoint instead of mountpoint as both
children and parent have it and they are relative.

Cherry-picked from Virtuozzo criu:
https://src.openvz.org/projects/OVZ/repos/criu/commits/e58a91883

Signed-off-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
---
 criu/mount.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/criu/mount.c b/criu/mount.c
index 3e907172d..516752a98 100644
--- a/criu/mount.c
+++ b/criu/mount.c
@@ -905,11 +905,11 @@ static int root_path_from_parent(struct mount_info *m, char *buf, int size)
 	bool head_slash = false, tail_slash = false;
 	int p_len, m_len, len;
 
-	if (!m->parent)
+	if (!m->parent || m->parent == root_yard_mp)
 		return -1;
 
-	p_len = strlen(m->parent->mountpoint);
-	m_len = strlen(m->mountpoint);
+	p_len = strlen(m->parent->ns_mountpoint);
+	m_len = strlen(m->ns_mountpoint);
 
 	len = snprintf(buf, size, "%s", m->parent->root);
 	if (len >= size)
@@ -925,11 +925,11 @@ static int root_path_from_parent(struct mount_info *m, char *buf, int size)
 	len = m_len - p_len;
 	BUG_ON(len < 0);
 	if (len) {
-		if (m->mountpoint[p_len] == '/')
+		if (m->ns_mountpoint[p_len] == '/')
 			head_slash = true;
 
 		len = snprintf(buf, size, "%s%s", (!tail_slash && !head_slash) ? "/" : "",
-			       m->mountpoint + p_len + (tail_slash && head_slash));
+			       m->ns_mountpoint + p_len + (tail_slash && head_slash));
 		if (len >= size)
 			return -1;
 	}
-- 
2.35.1

